File: my-folder/1046-max-consecutive-ones-iii/solution.java
class Solution {
    public int longestOnes(int[] nums, int k) {
		int len = 0;
		int start = 0;
		int max = 0;
		if (k == 0) {
			for (int i = 0; i < nums.length; i++) {
				if (nums[i] == 0) {
					len = 0;
				} else {
					if (++len > max) {
						max = len;
					}
				}
			}
			return max;
		}
		int flip = k;
		for (int i = 0; i < nums.length; i++) {
			if (nums[i] == 0) {
				if (flip == 0) {
					int j;
					for (j = start; j < i; j++) {
						len--;
						if (nums[j] == 0) {
							flip++;
							start = j + 1;
							break;
						}
					}
					if (flip == 0) {
						flip = 1;
						len = 0;
						start = i + 1;
						continue;
					}
				}
				flip--;
			}
			if (++len > max) {
				max = len;
			}
		}
		return max;
    }
}

File: my-folder/1046-max-consecutive-ones-iii/solution.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int l = 0;
        int r = 0;
        int cur = 0;
        int res = 0;
        while (r < nums.size())
        {
            cur += nums[r] == 0 ? 1 : 0;
            if (cur <= k)
            {
                res = max(res, r - l + 1);
                r++;
            }
            else
            {
                cur -= nums[l] == 0 ? 1 : 0;
                l++;
                r++;
            }
        }
        return res;
    }
};
